src/System.Management.Automation/engine/runtime/Binding/Binders.cs

```diff
@@ -4046,7 +4056,12 @@ internal static bool CanIndexFromEndWithNegativeIndex(DynamicMetaObject target)
             }
 
             // target implements IList<T>?
-            return limitType.GetInterfaces().Any(i => i.IsGenericType && i.GetGenericTypeDefinition() == typeof(IList<>));
+            if (limitType.GetInterfaces().Any(i => i.IsGenericType && i.GetGenericTypeDefinition() == typeof(IList<>)))
+            {
+                return true;
+            }
+
+            return typeof(ITuple).IsAssignableFrom(limitType);
```

Comment:
I wonder if this whole method shouldn't just check for a `Count` or `Length` property.

Reply:
I think that would be a breaking change.  For example, if a `Dictionary<,>` could make it that far into the binder (I know it can't, but custom code might) then `-1` could be a valid key.

I did a quick scan of the types visible by default in RC1, looking for types that:

1. Are decorated with `DefaultMemberAttribute`
1. Do not implement `ICollection` or `ICollection<>` (as the `Count` property coming from these make it likely enough to work as expected)
1. Do not implement `IDictionary<,>` (as they'll be handled earlier in the binder)

There's a couple types that jump out at a glance as *possibly* an issue, depending on how imperative it is to preserve that behavior. 

Here's the list in full, a couple obvious ones that jump out are `System.Linq.ILookup<,>`, `ReadOnlyPSMemberInfoCollection<>`, and some CIM related classes like `CimMethodParameterCollection`. 

```
System.String
System.ReadOnlySpan`1[T]
System.Span`1[T]
System.Numerics.Vector`1[T]
System.Text.StringBuilder
System.Runtime.CompilerServices.ITuple
System.Management.Automation.PSEventArgsCollection
System.Management.Automation.ReadOnlyPSMemberInfoCollection`1[T]
System.Management.Automation.Runspaces.InitialSessionStateEntryCollection`1[T]
System.Runtime.CompilerServices.IRuntimeVariables
Microsoft.Management.Infrastructure.CimMethodParametersCollection
Microsoft.Management.Infrastructure.Generic.CimKeyedCollection`1[T]
Microsoft.Management.Infrastructure.Generic.CimReadOnlyKeyedCollection`1[T]
System.Xml.XmlCDataSection
System.Xml.XmlCharacterData
System.Xml.XmlComment
System.Xml.XmlNodeList
System.Xml.XmlSignificantWhitespace
System.Xml.XmlText
System.Xml.XmlWhitespace
System.Xml.Serialization.XmlMembersMapping
System.Xml.Schema.XmlSchemaObjectTable
System.Collections.Specialized.StringDictionary
System.Linq.ILookup`2[TKey,TElement]
System.Linq.Lookup`2[TKey,TElement]
System.Data.SqlTypes.SqlChars
System.Data.SqlTypes.SqlBinary
System.Data.SqlTypes.SqlBytes
System.ComponentModel.MaskedTextProvider
Microsoft.CodeAnalysis.ChildSyntaxList
Microsoft.CodeAnalysis.SeparatedSyntaxList`1[TNode]
Microsoft.CodeAnalysis.SyntaxList`1[TNode]
Microsoft.CodeAnalysis.SyntaxNodeOrTokenList
Microsoft.CodeAnalysis.SyntaxTokenList
Microsoft.CodeAnalysis.SyntaxTriviaList
Microsoft.CodeAnalysis.Text.SourceText
Microsoft.CodeAnalysis.Text.TextLineCollection
System.Reflection.Metadata.GenericParameterHandleCollection
System.Reflection.Metadata.GenericParameterConstraintHandleCollection
Markdig.Helpers.StringSlice
```

While not incredibly likely to actually break existing code, perhaps it should be something the committee discusses?

Reply:
The risk is that an indexer supports negative values as an index somehow, e.g. as the key in a dictionary.

Based on your list, `StringDictionary`, ``Lookup`2`` and ``ILookup`2`` do look like they'd support a negative value as the index.

Maybe the check should be - the indexer accepts a single `int` and is not a generic parameter. This avoids the problem with `StringDictionary` with the signature `string Item(string key) {get;set;}` and `ILookup[int,T]]` where even though the indexer takes an `int`, it was generic and therefore not always an int.